### DISCUSSION TITLE
[help] How to use shapefiles via gadm with targets

### DISCUSSION THREAD

**Discussion by njtierney** [ropensci/targets#1240] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi there,

I've had some success with storing raster files as geotiffs in targets, as per the solution here: https://github.com/ropensci/targets/discussions/1238

However I cannot work out how to store a shapefile? I had assumed that I could write a `format_shapefile` format:

```r
format_shapefile <- tar_format(
  read = function(path) terra::vect(path),
  write = function(object, path) {
    terra::writeVector(
      x = object,
      filename = path,
      filetype = "ESRI Shapefile",
      overwrite = TRUE
    )
  },
  marshal = function(object) terra::wrap(object),
  unmarshal = function(object) terra::unwrap(object)
)
```

That could be used like so:

```r
tar_target(
    example_shapefile,
    get_example_shapefile(),
    format = format_shapefile
  )
```

Where `get_example_shapefile()` is shown by

```r
get_example_shapefile <- function() {

  f <- system.file("ex/lux.shp", package="terra")
  v <- vect(f)
  v

}
```

I've written a reprex here: https://github.com/njtierney/demo-geospatial-targets

**Comment by brownag** [comment, marked as answer]
I think the main issue with a shapefile is that it is actually multiple files.

`tar_format(..., write=...)` argument requires that `The function need not return a value, but the file written to path must be a single file, and it cannot be a directory.`

So, perhaps the simplest solution would be to use a different vector storage format that is only a single file. (I personally like .gpkg/GeoPackage for this). If each GeoPackage contains only one layer it is functionally equivalent to a shapefile without limitations of DBF etc.

If you want to make shapefile work, or any target that might constitute multiple related files, can use a trick I suggested in side discussion of https://github.com/ropensci/targets/discussions/809#discussioncomment-8461268 and bundle the multiple files in a ZIP file. As discussed in #809 there are analogous issues with GeoTIFF that contain categorical information, as this is stored in a .tif.aux.xml "sidecar" file by GDAL.

For zipped spatial data sources you can then use [GDAL virtual filesystem](https://gdal.org/user/virtual_file_systems.html) `"/vsizip/"` file path prefix in the `read` portion of tar_format(). This also works with various other compressed formats GDAL supports. The trick is in order for GDAL to recognize the path as a zipped source, you need to have the target name end in ".zip" (or equivalent) 

I updated your reprex [here](https://github.com/brownag/demo-geospatial-targets) with an example of how this can be done.

The bulk of the changes were in the `format_shapefile` portion, as well as adding `".zip"` suffix to shapefile target names in _targets.R:
```r
format_shapefile <- tar_format(
  read = function(path) terra::vect(paste0("/vsizip/", file.path(path, gsub("\\.zip", ".shp", basename(path))))),
  write = function(object, path) {
    terra::writeVector(
      x = object,
      filename = gsub("\\.zip", ".shp", basename(path)),
      filetype = "ESRI Shapefile",
      overwrite = TRUE
    )
    zf <- list.files(pattern = gsub("\\.zip", "", basename(path)))
    utils::zip(path, zf)
    unlink(zf)
  },
  marshal = function(object) terra::wrap(object),
  unmarshal = function(object) terra::unwrap(object)
)
```

When run it gives the following output running `tar_make()` in a clean instance, then running it again to verify nothing gets re-run. Calling `tar_load_everything()` then results in functional SpatVector objects equivalent to what we started with:
<details>

```

> tar_make()
Loading required package: terra
terra 1.7.73
▶ dispatched target country_codes
● completed target country_codes [0.008 seconds]
▶ dispatched target example_shapefile.zip
  adding: example_shapefile.cpg (stored 0%)
  adding: example_shapefile.dbf (deflated 78%)
  adding: example_shapefile.prj (deflated 17%)
 (deflated 50%)le_shapefile.shp
  adding: example_shapefile.shx (deflated 31%)
● completed target example_shapefile.zip [0.021 seconds]
▶ dispatched target example_rast
● completed target example_rast [0.01 seconds]
▶ dispatched target example_gadm.zip
trying URL 'https://geodata.ucdavis.edu/gadm/gadm4.1/pck/gadm41_AUS_0_pk_low.rds'
Content type 'unknown' length 701230 bytes (684 KB)
==================================================
downloaded 684 KB

trying URL 'https://geodata.ucdavis.edu/gadm/gadm4.1/pck/gadm41_CXR_0_pk_low.rds'
Content type 'unknown' length 1615 bytes
==================================================
downloaded 1615 bytes

trying URL 'https://geodata.ucdavis.edu/gadm/gadm4.1/pck/gadm41_CCK_0_pk_low.rds'
Content type 'unknown' length 4205 bytes
==================================================
downloaded 4205 bytes

trying URL 'https://geodata.ucdavis.edu/gadm/gadm4.1/pck/gadm41_HMD_0_pk_low.rds'
Content type 'unknown' length 3046 bytes
==================================================
downloaded 3046 bytes

trying URL 'https://geodata.ucdavis.edu/gadm/gadm4.1/pck/gadm41_NZL_0_pk_low.rds'
Content type 'unknown' length 225249 bytes (219 KB)
===========
downloaded 219 KB

  adding: example_gadm.cpg (stored 0%)
  adding: example_gadm.dbf (deflated 59%)
  adding: example_gadm.prj (deflated 17%)
  adding: example_gadm.shp (deflated 61%)
  adding: example_gadm.shx (deflated 35%)
trying URL 'https://geodata.ucdavis.edu/gadm/gadm4.1/pck/gadm41_NFK_0_pk_low.rds'
Content type 'unknown' length 1570 bytes
==================================================
downloaded 1570 bytes

● completed target example_gadm.zip [5.485 seconds]
▶ completed pipeline [5.904 seconds]
 
> tar_make()
Loading required package: terra
terra 1.7.73
✔ skipped target country_codes
✔ skipped target example_shapefile.zip
✔ skipped target example_rast
✔ skipped target example_gadm.zip
✔ skipped pipeline [0.116 seconds]

> tar_load_everything()

```

</details>

**Reply by njtierney** [reply]
Thanks so much @brownag ! Really appreciate your extensive example.

I've cloned your repo and it works great! But even copying your `format_shapefile` and relevant code, I run into this error:

```
▶ dispatched target country_codes
● completed target country_codes [0.004 seconds]
▶ dispatched target example_gadm
✖ errored target example_gadm
✖ errored pipeline [0.129 seconds]
Warning message:
Layer 'example_gadm' already exists (GDAL error 1) 
Error:
! Error running targets::tar_make()
Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
Debugging guide: https://books.ropensci.org/targets/debugging.html
How to ask for help: https://books.ropensci.org/targets/help.html
Last error message:
    _store_ [writeVector] Layer creation failed
Last error traceback:
    No traceback available.
```

**Reply by njtierney** [reply]
I've got a PR with those changes (plus some others) - https://github.com/njtierney/demo-geospatial-targets/pull/2 - I'm wondering if you have any thoughts as to why I don't get the same answer? If you could make a PR from your repo to mine I could see a diff and that might also provide some clues :)

**Reply by brownag** [reply]
I put some suggested changes in your PR in order to get your branch to run. The main thing is the ".zip" suffix on the `example_gadm` target name is needed for both write and read... For write so it goes into the base directory (code currently relies on that to find and cleanup the file), and for read such the target object stored can be recognized as a ZIP source by GDAL

**Reply by njtierney** [reply]
Thanks so much!

Ah I was not anticipating the `.zip` suffix to be required - should have checked that!

I can't seem to see changes at https://github.com/njtierney/demo-geospatial-targets/pull/2 - did I miss something?

**Reply by brownag** [reply]
I did it as a review on the PR with the comment-style suggested changes. I think you need to approve the review / accept the changes in order for them to be incorporated.

https://github.com/njtierney/demo-geospatial-targets/pull/2/files/ff91675e8c3c14fc19d37daa60182934447d6893

**Reply by brownag** [reply]
Oops, sorry, the review was not completed on my end. You should be able to see them now

**Reply by njtierney** [reply]
Ah awesome, thanks so much! Definitely feels like there should be a targets package for spatial data, I reckon this could be really useful for other folks.

**Reply by njtierney** [reply]
That all worked perfectly for me, thank you so much @brownag ! :)
